source/_posts/2015-03-26-creating-a-dynamic-single-page-app-for-our-genome-team-using-react.markdown

```diff
@@ -0,0 +1,96 @@
+ 
+##### Drag and drop
+To implement drag and drop in React, you have to both move the items around visually and change the state of the page every time something changes. Because we were moving around large components, we basically had to rerender the entire page when a component moved. In order to avoid excessive page lag, we wrote our own drag/drop functions using native HTML5 listeners.
+
+Our solution was to store the dragged element and starting mouse position in the panel's state, and to swap the elements when the dragging element != the element you are currently hovering over. We use the change in mouse position to determine if you are moving up or down and reorder the page elements accordingly.
+
+A snippet of our drag/drop code:
+```
+  dragStart: function(e) {
+    this.setState({dragMouseStart: e.clientY, draggingElementId: e.currentTarget.getAttribute('data-slug')})
+    e.dataTransfer.effectAllowed = 'move';
+  },
+  dragOver: function(e) {
+    e.preventDefault();
+
+    var over = e.currentTarget; // element we're hovering over
+    var draggingId = this.state.draggingElementId;
+    var toElId = over.getAttribute('data-slug');
+    var movingUp = (this.state.dragMouseStart > e.clientY);
+
+    if (draggingId != toElId) {
+      orderCopy = this.state.artworksOrder
+      var toIndex = orderCopy.indexOf(toElId)
+      var fromIndex = orderCopy.indexOf(draggingId)
+      if (movingUp) {
+        // add a comment for what this clever line does?
+        orderCopy.splice(toIndex, 0, orderCopy.splice(fromIndex, 1)[0])
+      } else {
+        orderCopy.splice(fromIndex, 0, orderCopy.splice(toIndex, 1)[0])
+      }
+      this.setState({artworksOrder: orderCopy})
+    }
+  }
+```
+ 
+### Writing Specs
+
+All of the existing specs for the app were written in Rspec, so we chose to write integration tests using Rspec+Capybara. The headless Capybara webkit did not integrate with our React setup, so we switched to Selenium (which also conveniently let us debug our specs within the browser).
+ 
+Our main challenge with specs had to do with Rspec not waiting long enough for components (such as autocomplete results) to appear, perhaps due to the fake React-DOM. We spent many sad hours debugging mysteriously spurious tests, and even including a few dreaded 'sleep' commands. Eventually, we integrated the [rspec-retry](https://github.com/y310/rspec-retry) gem to retry spurious tests during CI.
+
+### Conclusion
+ 
+Converting our app to have a React-based front end was surprisingly simple. We were able to incrementally change certain templates to React components, which made it easy to test as we went along. Additionally, our development time (and emotional state) to add new features since then has decreased dramatically. It is much easier to add new components or edit existing ones when there is a single source of truth and you don't have to search through semantically unorganized coffeescript files.
```

Comment:
>Additionally, our development time (and emotional state) to add new features since then has decreased dramatically.

a state can't decrease dramatically.  maybe "emotional cost" instead?  not sure.